x-pack/plugin/vectors/src/main/java/org/elasticsearch/xpack/vectors/mapper/DenseVectorLSHFieldMapper.java

```diff
@@ -0,0 +1,393 @@
+            return null;
+        }
+
+        @Override
+        public Query termQuery(Object value, QueryShardContext context) {
+            return new TermQuery(new Term(name(), (BytesRef) value));
+        }
+
+        int dims() {
+            return dims;
+        }
+
+        void setDims(int dims) {
+            this.dims = dims;
+        }
+
+        public LSHModelEuclideanMultiProbe lshModel() {
+            return lshModel;
+        }
+
+        void setLSHModel(LSHModelEuclideanMultiProbe lshModel) {
+            this.lshModel = lshModel;
+        }
+    }
+
+    protected DenseVectorLSHFieldMapper(String simpleName, MappedFieldType fieldType, MappedFieldType defaultFieldType,
+                                        Settings indexSettings, MultiFields multiFields, CopyTo copyTo) {
+        super(simpleName, fieldType, defaultFieldType, indexSettings, multiFields, copyTo);
+    }
+
+    @Override
+    protected DenseVectorLSHFieldMapper clone() {
+        return (DenseVectorLSHFieldMapper) super.clone();
+    }
+
+    @Override
+    public DenseVectorLSHFieldType fieldType() {
+        return (DenseVectorLSHFieldType) super.fieldType();
+    }
+
+    @Override
+    public void parse(ParseContext context) throws IOException {
+        // TODO: implement usage in multi-fields
+        if (context.externalValueSet()) {
+            throw new IllegalArgumentException("Field [" + name() + "] of type [" + typeName() + "] can't be used in multi-fields");
+        }
+
+        int dims = fieldType().lshModel.dims(); // number of vector dimensions
+        float[] vector = new float[dims];
+        int dim = 0;
+        for (Token token = context.parser().nextToken(); token != Token.END_ARRAY; token = context.parser().nextToken()) {
+            if (dim >= dims) {
+                throw new IllegalArgumentException("Field [" + name() + "] of type [" + typeName() + "] of doc [" + context.doc() +
+                    "] has number of dimensions [" + dim + "] greater than defined in the mapping: [" +  dims +"]");
+            }
+            ensureExpectedToken(Token.VALUE_NUMBER, token, context.parser()::getTokenLocation);
+            vector[dim] = context.parser().floatValue(true);
+            dim++;
+        }
+        if (dim != dims) {
+            throw new IllegalArgumentException("Field [" + name() + "] of type [" + typeName() + "] of doc [" + context.doc() +
+                "] has number of dimensions [" + dim + "] less than defined in the mapping: [" +  dims +"]");
+        }
+
+        // hash vector to get L number of hashes corresponding to L hash tables
+        BytesRef[] hashes = fieldType().lshModel().hash(vector);
+        // index each hash as an untokenized term
+        String fieldNameLSH = fieldType().name();
+        for (int i = 0; i < hashes.length; i++) {
+            Field indexedField = new Field(fieldNameLSH, hashes[i], fieldType());
```

Comment:
Is there a concern over hash collisions? I was expecting that each hash table would correspond to a different lucene field.

Reply:
Thanks, Julie. There should not be hash collisions for different hash tables. When we encode a hash into a Lucene term, [first two bytes  ](https://github.com/elastic/elasticsearch/pull/44374/files#diff-0130915d8e2beff968c329518d664a21R247) we allocate for the hash table number.  This means that the terms belonging to different hash tables will never match or collide.

Reply:
Thanks for clarifying, I hadn't read the encoding logic carefully yet.